### DISCUSSION TITLE
Data aggregation

### DISCUSSION THREAD

**Discussion by arif-hanif** [ChilliCream/hotchocolate#2963] [Ideas]
@PascalSenn started the proposal, we can discuss spec here.
https://github.com/ChilliCream/hotchocolate/issues/924

Who all is interested in data aggregation for different data sources?
Entity Framework
MongoDb
Neo4j

# Aggregations
> This proposal is very **Work In Progress**
## Introduction
Aggregations are very useful for querying big data sources. To count items or sum a value, it's not necessary to load all the data. Querying already aggregated data in the backend has less overhead. 

## Proposal
A new API Endpoint is generated for a collection. The type of this endpoint has different aggregation fields on them. 

It will not be possible to cover all aggregation cases. We simply cannot create types for all possible scenarios. 

Proposed are the following: 

### Example Objects
All examples will take place in the following domain:
```csharp
public class Foo {
     public string role {get;set;}
     public int num {get;set;}
     public Bar bars {get;set;}
}

public class Bar {
     public string baz {get;set;}
     public int blub {get;set;}
}
```

### count 
```graphql
{
  foosMeta {
    count
}
```
returns the amount of foo's in foos.

### sum
```graphql
{
  foosMeta {
    sum_num
}
```
returns the sum of all nums.

### average
```graphql
{
  foosMeta {
    sum_average
}
```
returns the average of all nums.

### Example
```graphql
{
  foosMeta {
    group_by {
      role {
        key
        value {
          bar {
            baz
          }
        }
      }
    }
  }
}
```

#### Prisma (Issue https://github.com/prisma/prisma/issues/1312) 
```graphql 
# Count all posts with a title containing 'GraphQL'
query {
  postsConnection(where: {
    title_contains: "GraphQL"
  }) {
    aggregate {
      count
    }
  }
}
{
  allUsersConnection(where: {city: "Aarhus"}) {
    aggregate {
      avg {
        age
      }
      max {
        age
      }
    }
    edges {
      node { name, age }
    }
  }
}
{
  allUsersConnection(where: {city: "Aarhus"}) {
    aggregate {
      avg {
        age
      }
      max {
        age
      }
    }
    edges {
      node { name, age }
    }
  }
}
```

#### Horsura 
```graphql
query {
  article_aggregate {
    aggregate {
      count
      sum {
        rating
      }
      avg {
        rating
      }
      max {
        rating
      }
    }
    nodes {
      id
      title
      rating
    }
  }
}
query {
  author (where: {id: {_eq: 1}}) {
    id
    name
    articles_aggregate {
      aggregate {
        count
        avg {
          rating
        }
        max {
          rating
        }
      }
      nodes {
        id
        title
        rating
      }
    }
  }
}
```

#### Gatsby 
```graphql
### no example found
```

#### GraphCMS
```graphql
### no example found
```

**Comment by michaelstaib** [comment]
I think we should follow Prisma and aggregate on the connection.

**Reply by michaelstaib** [reply]
This essentially would work with both the offset paging and the cursor paging.

**Reply by twilly86** [reply]
I want the aggregate to sum on all records outside paging. Would this be possible?  So if I have page limit 10 and 1000 records the count would be 1000

**Comment by a5ter0id** [comment]
@michaelstaib 
1. Could you tell us the upcoming plan or maybe the release date of the aggregation/grouping functionality?
2. Do you have a source code (alpha version) that implements this functionality. Maybe we can cooperate/collaborate and develop it together?

**Comment by MaheshB0ngani** [comment]
This is all I am waiting for. Do you have any roadmap for this. If there is any initial work done on this please share the branch details so that we can understand the details and complexity, thus we can also contribute parallelly.

This is what I am looking for [link](https://github.com/ChilliCream/hotchocolate/issues/924#issuecomment-975777556)

**Comment by mu-dawood** [comment]
any update on that

**Reply by a5ter0id** [reply]
I see that the plan is 2022
https://github.com/ChilliCream/hotchocolate/projects/28#card-55772065

but I am not sure that something will happen this year
because this date is updated several times and no progress yet

**Reply by michaelstaib** [reply]
It's in the 2022 backlog, which does not mean it will be pulled. At the moment we are focusing on papercuts and schema-stitching.

**Comment by aaron-manning** [comment]
This feature is pretty vital to bridge the feature gap between HC and hasura / postgraphile.

**Comment by M2xAy** [comment]
Hi Community
When will you release aggregation on nodes?
Best regards

**Comment by ramagott** [comment]
Hi @michaelstaib 

We are also planning to use this data aggregation feature in our service. Is there any update on the timeline?

**Comment by andreamorello93** [comment]
hi @michaelstaib any update on this?

**Comment by horia-calborean** [comment]
This would help a lot. You can think about exposing GraphQL to an AI assistant. Having these tools out of the box would be a huge benefit
